README.md
# rnglib

A Python random number generator especially useful for generating
quasi-random strings, data files, and directory trees containing
random numbers of data files with random names and random content.

## Classes

This package contains three classes, **SimpleRNG**, **SystemRNG**,
and **SecureRNG**.  All of these subclass Python's `random` standard
library and so `random`'s functions can be called through any
of `rnglib`'s three subclasses.

**SimpleRNG** is the fastest of the subclasses.  It uses the
[Mersenne Twister](https://en.wikipedia.org/wiki/Mersenne_Twister)
and so is completely predictable with a very long period.
It is suitable where speed and predictability are both important.
That is, you can be certain that if you provide the same
[seed](https://en.wikipedia.org/wiki/Random_seed),
then the sequence of numbers generated will be exactly the same.  This is
often important for debugging.

**SystemRNG** is a secure random number generator, meaning that it is
extremely difficult or impossible to predict the next number
generated in a sequence.  It is based on the system's `/dev/urandom`.
This relies in part upon
[entropy](https://en.wikipedia.org/wiki/Entropy (Computing))
accumulated by the system; when
this is insufficient (when we run out of entropy) it will revert to using
a very secure programmable random number generator.

**SecureRNG** is a very secure random number generator based
`/dev/random`.  This also relies upon entropy accumulated by the
system, but will **block** when there is not enough entropy available
to generate the random value requested.  This means that it can be
very slow.

For normal use in testing, `SimpleRNG` is preferred.  When there is
a requirement for a reasonable level of security, as in the
generation of passwords and cryptographic keys with reasonable
strength, `SystemRNG` is recommended.  Where there is a need for
very secure random values, use `SecureRNG` but expect to wait a
while for results.

## Functions

In addition to the functions available from Python's `random` package,
`rnglib` also provides

	next_boolean()
	next_byte(max=256)
	next_bytes(buffer)
	next_int16(max=65536])
	next_int32(max=65536*65536)
	next_int64(max=65536*65536*65536*65536)
	next_real()
	next_file_name(max_len)
	next_data_file(dir_name, max_len, min_len=0)
	next_data_dir(pathToDir, depth, width, max_len, min_len=0)

Given a buffer of arbitrary length, `next_bytes()` will fill it with random
bytes using the chosen random number generator (`SimpleRNG`, `SystemRNG`,
or `SecureRNG`).

File names generated by `next_file_name()` are at least one character long.
The first letter must be alphabetic (including the underscore)
whereas other characters may also be numeric or either of dot and dash
("." and "-").

The function `next_data_file()` creates a file with a random name in the
directory indicated.  The file length will vary from min_len (which
defaults to zero) up to but excluding max_len bytes.

The `next_data_dir()` function creates a directory structure that is
'depth' deep and 'width' wide, where the latter means that there
will be that many files in each directory and subdirectory created
(where a file is either a data file or a directory). `depth` and
`width` must both be at least 1.  `max_len` and `min_len` are used as in
`next_data_file()`.

## History

`rnglib` was implemented as part of the
[XLattice](http://www.xlattice.org)
project.  A Go language implementation forms part of
[xlattice_go.](https://github.com/jddixon/xlattice_go)

## Project Status

Stable, good benchmarks.


## On-line Documentation

More information on the **rnglib** project can be found
[here](https://jddixon.github.io/rnglib)


CHANGES
~/dev/py/rnglib/CHANGES

v1.3.10
    2019-03-19
        * cosmetic changes
v1.3.9
    2018-03-20
        * fix corrupt .dvcz/builds                                  SLOC 1483
v1.3.8
    2018-02-26
        * correct config files                                      SLOC 1478
v1.3.7
    2017-11-28
        * pep8; made config files consistent                        SLOC 1485
v1.3.6
    2017-09-23
        * corrections to setup.py, move projConfig.toml             SLOC 1485
v1.3.5
    2017-09-20
        * add projConfig.toml                                       SLOC 1465
v1.3.4
    2017-09-12
        * pep8 etc; fix tox.ini                                     SLOC 1436
v1.3.3
    2017-07-27
        * rewrite setup.py; minor edits to config files             SLOC 1410
v1.3.2
    2017-06-22
        * restructure for pytest                                    SLOC 1405
v1.3.1
    2017-05-04
        * drop dist/* from repos; plus minor edits                  SLOC 1475
v1.3.0
    2017-03-30
        * fix minor bugs, correct SecureRNG and test                SLOC 1475
v1.2.1
    2017-02-19
        * comment out deprecated code                               SLOC 1391
v1.2.0
    2017-01-13
        * rename nextByte to next_byte throughout
        * applied PEP 8 to function/method names in README.md
        * corrected raw sring FILE_NAME_CHARS to exclude backslash  SLOC 1199
v1.1.3
    2017-01-03
        * correct errors in valid_file_name; fix tests to catch     SLOC 1185
v1.1.2
    2016-12-29
        * pylinting, minor bug fixes                                SLOC 1179
v1.1.1
    2016-12-20
        * add valid_file_name()                                     SLOC 1388
v1.1.0
    2016-10-08
        * minor cleanup                                             SLOC 1354
    2016-10-06
        * add pylint; refactor (forgot to step version nbr)         SLOC 1351
v1.0.14
    2016-08-30
        * improve coverage from 79% to 94%                          SLOC 1105
v1.0.13
    2016-08-27
        * fix setup.py, license                                     SLOC 1132
v1.0.12
    2016-06-28
        * add ghpDoc/{css,img}                                      SLOC 1024
v1.0.11
    2016-05-31
        * add bkp2U, other bash scripts; update .dvcz               SLOC  565
v1.0.10
    2016-05-03
        * add .dvcz and projCfg*                                    SLOC  526
v1.0.9
    2016-03-27
        * add and execute auto8it                                   SLOC  484
v1.0.8
    2016-02-25
        * update LICENSE                                            SLOC  470
v1.0.7
    2016-01-31
        * minor edits                                               SLOC  470
v1.0.6
    2015-12-26
        * fix links, other minor edits
v1.0.5
    2015-11-17
        * minor edits to *.md                                       SLOC  471
v1.0.4
    2015-09-12
        * add refs, improve formatting of docs                      SLOC  471
v1.0.3
    2015-08-10
        * minor cleanup; correct copyright dates, etc               SLOC  471
v1.0.2
    2015-06-13
        * add installit, update .gitignore                          SLOC  471
v1.0.1
    2015-05-18
        * verify that two SimpleRNGs with the same seed generate    SLOC  470
            the same sequences  
v1.0.0
    2014-12-13
        * convert to python3                                        SLOC  424
v0.7.5
    2014-08-25
        * added countem, test of someBytes                          SLOC  427
v0.7.4
    2014-06-3
        * added someBytes(N) which returns a bytearray of that 
            many random values
v0.7.3
    2013-06-16
        * corrected default salt for SimpleRNG etc                  SLOC  406

v0.7.2
    2013-06-10
        * modified nextBytes(buffer) to use random.getrandombits    SLOC  406
            on shorter arrays and os.urandom on longer ones,
            switching from one to the other at a point experiments
            suggest is optimal
v0.7.1
    2013-06-06
        * changed SimpleRNG.__init__ to follow random.Random, so 
            param is (a=None) and __init__ then invokes 
            random.salt(a), which if a is None uses either
            urandom() or time.time()
        * added testRandomFunc.py, which exercises the              SLOC  399
            random.Random functions available through SimpleRNG
v0.7.0
    2013-06-05
        * dropped AbstractRNG, redefined SimpleRNG so that it 
            subclasses random.Random, added SystemRNG which 
            subclasses random.SystemRandom
v0.6.3
    2013-06-04
        * tests create tmp/ if it does not exist
v0.6.2 
    2012-06-01
        * corrected minor fault in SimpleRNG.nextDataFile(): now 
            confirms that a random file name is not in use before 
            creating the file
v0.6.1
    2012-05-30
        * removed rnglib.py (which has been merged into 
            rnglib/__init__.py)
        * moved dist/myrng* into dist.old.names/
0.6.0
    2012-04-24 jdd
        * renamed package to rnglib 


0.5 
    2011-09-04 
        * jdd replaced MAX_INT16 and such with numeric values
                     (65536 and such) because compiler didn't like symbols

0.3 
    2011-08-11 
        * jdd restructured to make rnglib.py a module

0.2 
    2011-07-12 
        *changed nextFileName() to not produce '.'
0.1 
    2011-07-11 
        * passes tests






























test_files.py
#!/usr/bin/env python3
# testDataFile.py

""" Exercise DataFile functionality. """

import unittest

import rnglib


class TestDataFile(unittest.TestCase):
    """ Exercise DataFile functionality. """

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_simplest_constructor(self):
        """ Verify that the constructor works correctly. """

        leaf0 = rnglib.DataFile("foo")
        self.assertEqual("foo", leaf0.name)
        self.assertEqual(None, leaf0.parent)
        self.assertEqual(leaf0.name, leaf0.path)

        leaf1 = rnglib.DataFile("bar")
        self.assertEqual("bar", leaf1.name)

        self.assertTrue(leaf0.__eq__(leaf0))
        self.assertEqual(leaf0, leaf0)
        self.assertFalse(leaf0.__eq__(leaf1))


if __name__ == '__main__':
    unittest.main()


test_simple_rng.py
#!/usr/bin/env python3
# testSimpleRNG.py

""" Test the simple random number generator. """

import os
import shutil
import time
import unittest
import rnglib

TEST_DIR = 'tmp'


class TestSimpleRNG(unittest.TestCase):
    """ Test the simple random number generator. """

    def setUp(self):
        now = time.time()
        self.rng = rnglib.SimpleRNG(now)
        os.makedirs(TEST_DIR, exist_ok=True, mode=0o755)

    def tearDown(self):
        pass

    # utility functions #############################################
    def _build_data(self, count):
        """ Return the requested number of quasi-random bytes. """

        self.assertTrue(count > 0)
        data = bytearray(count)
        for i in range(count):
            self.assertTrue(data[i] == 0)
        self.assertEqual(count, len(data))
        return data

#   # XXX NOT CURRENTLY USED
#   def set_a_bit(self, vector, value):
#       """ treat a 32 byte vector as a bit vector of 256 bits """
#       byte = int(value / 8)
#       bit = value % 8
#       vector[byte] |= 1 << bit
#       return vector

#   # XXX NOT CURRENTLY USED
#   def non_zero_bits(self, vector):
#       pass

    # actual unit tests #############################################
    def test_constants(self):
        """" Verify constants have expected values. """

        max_int16 = rnglib.MAX_INT16
        max_int32 = rnglib.MAX_INT32
        max_int64 = rnglib.MAX_INT64

        self.assertEqual(65536, max_int16)
        self.assertEqual(max_int16 * max_int16, max_int32)
        self.assertEqual(max_int32 * max_int32, max_int64)

    def test_simplest_constructor(self):
        """ Verify that the class's random number generator is not None. """

        self.assertFalse(self.rng is None)

    def test_seed(self):
        """ Check the bahavior of seed values. """

        seed = self.rng.next_int16()

        # if the seed is the same, the numbers generated should be the same
        rng1 = rnglib.SimpleRNG(seed)
        rng2 = rnglib.SimpleRNG(seed)

        for _ in range(16):
            aaa = rng1.next_int16()
            bbb = rng2.next_int16()
            self.assertEqual(aaa, bbb)

        # if the seeds differ, with a very high probability the numbers
        # generated should differ
        seed1 = (seed << 16) | seed
        seed2 = ~seed1
        rng1 = rnglib.SimpleRNG(seed1)
        rng2 = rnglib.SimpleRNG(seed2)
        aaa = rng1.next_int16()
        bbb = rng2.next_int16()
        self.assertTrue(aaa != bbb)     # fails, rarely

    def test_next_boolean(self):
        """ Check the behavior of the next_boolean() function. """

        value = self.rng.next_boolean()
        self.assertTrue((value is True) or (value is False))
        self.assertTrue(isinstance(value, bool))

    def test_next_byte(self):
        """ Check the behavior of the next_byte() function. """

        value = self.rng.next_byte()
        self.assertTrue((value >= 0) and (value < 256))
        # would like to test that the entire range is filled
        # where the cost of the test is reasonable

    def test_next_bytes(self):
        """ Check the behavior of the next_bytes() function. """

        length = 16 + self.rng.next_byte()
        self.assertTrue((length >= 16) and (length < 272))
        data = self._build_data(length)      # builds a byte array
        self.rng.next_bytes(data)
        self.assertEqual(length, len(data))

    def test_next_file_name(self):
        """ Check the behavior of the next_file() function. """

        for count in range(8):
            max_len = 16 + count
            name = self.rng.next_file_name(max_len)
            self.assertTrue(max_len > len(name))
            self.assertTrue(len(name) > 0)

    def test_next_data_file(self):
        """ Check the behavior of the next_data_file() function. """

        for _ in range(9):
            file_len = 16 + self.rng.next_byte()
            (count, path_to_file) = self.rng.next_data_file(
                TEST_DIR, file_len + 1, file_len)
            self.assertTrue(os.path.exists(path_to_file))
            self.assertEqual(os.path.getsize(path_to_file), count)

    def test_some_bytes(self):
        """ Check the behavior of the some_bytes() function. """

        now = time.time()
        rng = rnglib.SimpleRNG(now)
        for _ in range(8):
            count = 1 + rng.next_int16(16)
            b_val = rng.some_bytes(count)
            self.assertEqual(len(b_val), count)
            self.assertTrue(isinstance(b_val, bytearray))

    def do_next_data_dir_test(self, width, depth):
        """ Check the behavior of the next_data_dir() function. """

        dir_name = self.rng.next_file_name(8)
        dir_path = "%s/%s" % (TEST_DIR, dir_name)
        if os.path.exists(dir_path):
            if os.path.isfile(dir_path):
                os.unlink(dir_path)
            else:
                shutil.rmtree(dir_path)
        self.rng.next_data_dir(dir_path, width, depth, 32)

    def test_next_data_dir(self):
        """ Check the behavior of the next_data_dir() function. """

        self.do_next_data_dir_test(1, 1)
        self.do_next_data_dir_test(1, 4)
        self.do_next_data_dir_test(4, 1)
        self.do_next_data_dir_test(4, 4)


if __name__ == '__main__':
    unittest.main()


test_system_rng.py
#!/usr/bin/env python3
# testSystemRNG.py

""" Test the crypto-grade system random number generator. """

import os
import shutil
import time
import unittest
import rnglib

TEST_DIR = 'tmp'


class TestSystemRNG(unittest.TestCase):
    """ Test the crypto-grade system random number generator. """

    def setUp(self):
        now = time.time()
        self.rng = rnglib.SystemRNG(now)
        os.makedirs(TEST_DIR, exist_ok=True, mode=0o755)

    def tearDown(self):
        pass

    # utility functions #############################################
    def _build_data(self, count):
        self.assertTrue(count > 0)
        data = bytearray(count)
        for ndx in range(count):
            self.assertTrue(data[ndx] == 0)
        self.assertEqual(count, len(data))
        return data

#   # XXX NOT CURRENTLY USED
#   def set_a_bit(self, vector, value):
#       """ treat a 32 byte vector as a bit vector of 256 bits """
#       byte = int(value / 8)
#       bit = value % 8
#       vector[byte] |= 1 << bit
#       return vector

#   # XXX NOT CURRENTLY USED
#   def non_zero_bits(self, vector):
#       pass

    def test_constants(self):
        """ Verify constants have expected values. """
        max_int16 = rnglib.MAX_INT16
        max_int32 = rnglib.MAX_INT32
        max_int64 = rnglib.MAX_INT64

        self.assertEqual(65536, max_int16)
        self.assertEqual(max_int16 * max_int16, max_int32)
        self.assertEqual(max_int32 * max_int32, max_int64)

    def test_simplest_constructor(self):
        """ Verify local random number generator has been defined. """
        self.assertFalse(self.rng is None)

#   def test_seed(self):
#       pass

    def test_next_boolean(self):
        """ Test the secure next_boolean() function. """

        value = self.rng.next_boolean()
        self.assertTrue((value is True) or (value is False))
        self.assertTrue(isinstance(value, bool))

    def test_next_byte(self):
        """ Test the secure next_byte() function. """
        value = self.rng.next_byte()

        self.assertTrue((value >= 0) and (value < 256))
        # would like to test that the entire range is filled
        # where the cost of the test is reasonable

    def test_next_bytes(self):
        """ Test the secure next_bytes() function. """
        length = 16 + self.rng.next_byte()
        self.assertTrue((length >= 16) and (length < 272))
        data = self._build_data(length)      # builds a byte array
        self.rng.next_bytes(data)
        self.assertEqual(length, len(data))

    def test_next_file_name(self):
        """ Test the secure next_file_name() function. """
        for count in range(8):
            max_len = 16 + count
            name = self.rng.next_file_name(max_len)
            self.assertTrue(max_len > len(name))
            self.assertTrue(len(name) > 0)

    def test_next_data_file(self):
        """ Test the secure next_data_file() function. """
        for _ in range(9):
            file_len = 16 + self.rng.next_byte()
            (count, path_to_file) = self.rng.next_data_file(
                TEST_DIR, file_len + 1, file_len)
            self.assertTrue(os.path.exists(path_to_file))
            self.assertEqual(os.path.getsize(path_to_file), count)

    def do_next_data_dir_test(self, width, depth):
        """
        Test creating test data directory with these parameters.
        """
        dir_name = self.rng.next_file_name(8)
        dir_path = "%s/%s" % (TEST_DIR, dir_name)
        if os.path.exists(dir_path):
            if os.path.isfile(dir_path):
                os.unlink(dir_path)
            else:
                shutil.rmtree(dir_path)
        self.rng.next_data_dir(dir_path, width, depth, 32)

    def test_next_data_dir(self):
        """
        Test creating test data directories with various parameters.
        """
        self.do_next_data_dir_test(1, 1)
        self.do_next_data_dir_test(1, 4)
        self.do_next_data_dir_test(4, 1)
        self.do_next_data_dir_test(4, 4)

    def test_bulk_generation(self):
        """
        Test generating many AES iv+keys.
        """
        t0 = time.time()
        for i in range(10000):
            junk = self.rng.some_bytes(16)
        t1 = time.time()
        print("time to generate 10,000 AES iv + key: %f seconds" % (t1 - t0))


if __name__ == '__main__':
    unittest.main()


testFiles.py
#!/usr/bin/env python3

# testDataFile.py
import unittest

import rnglib


class TestDataFile (unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    # utility functions #############################################

    # actual unit tests #############################################
    def testSimplestConstructor(self):
        leaf0 = rnglib.DataFile("foo")
        self.assertEqual("foo", leaf0.name)
        self.assertEqual(None, leaf0.parent)
        self.assertEqual(leaf0.name, leaf0.path)

        leaf1 = rnglib.DataFile("bar")
        self.assertEqual("bar", leaf1.name)

        self.assertTrue(leaf0.__eq__(leaf0))
        self.assertEqual(leaf0, leaf0)
        self.assertFalse(leaf0.__eq__(leaf1))

if __name__ == '__main__':
    unittest.main()


testRandomFunc.py
#!/usr/bin/env python3

# ~/dev/py/rnglib/testRandomFunc.py

import math
import random
import time
import unittest

from rnglib import SimpleRNG

# This is not a test in the usual sense.  It exercises random.Random
# functions through a SimpleRNG instance and makes the results available
# for human inspection.

# This code is hacked from python2.7/random.py and so made available
# under the same license as Python itself.


class TestRandomFunc(unittest.TestCase):

    def doTest(self, n, func, args):
        print("%u invocations of %s" % (n, func.__name__))
        total = 0.0
        sqsum = 0.0
        smallest = 1e10
        largest = -1e10
        t0 = time.time()
        for i in range(n):
            x = func(*args)
            total += x
            sqsum = sqsum + x * x
            smallest = min(x, smallest)
            largest = max(x, largest)
        t1 = time.time()
        print("  %6.4f sec, " % round(t1 - t0, 4), end=' ')
        avg = total / n
        stddev = math.sqrt(sqsum / n - avg * avg)
        print('avg %g, stddev %g, min %g, max %g' %
              (avg, stddev, smallest, largest))

    def randTest(self, n=1000):
        rng = SimpleRNG(time.time())
        self.doTest(n, rng.random, ())
        self.doTest(n, rng.normalvariate, (0.0, 1.0))
        self.doTest(n, rng.lognormvariate, (0.0, 1.0))
        self.doTest(n, rng.vonmisesvariate, (0.0, 1.0))
        self.doTest(n, rng.gammavariate, (0.01, 1.0))
        self.doTest(n, rng.gammavariate, (0.1, 1.0))
        self.doTest(n, rng.gammavariate, (0.1, 2.0))
        self.doTest(n, rng.gammavariate, (0.5, 1.0))
        self.doTest(n, rng.gammavariate, (0.9, 1.0))
        self.doTest(n, rng.gammavariate, (1.0, 1.0))
        self.doTest(n, rng.gammavariate, (2.0, 1.0))
        self.doTest(n, rng.gammavariate, (20.0, 1.0))
        self.doTest(n, rng.gammavariate, (200.0, 1.0))
        self.doTest(n, rng.gauss, (0.0, 1.0))
        self.doTest(n, rng.betavariate, (3.0, 3.0))
        self.doTest(n, rng.triangular, (0.0, 1.0, 1.0 / 3.0))

    def testRandTest(self):
        self.randTest(n=1000)

if __name__ == '__main__':
    unittest.main()


src/rnglib/__init__.py
# rnglib/__init__.py

"""
See docs.python.org/2/library/random.html for advice on subclassing
random.Random.  Override random(), seed(), getstate(), setstate(),
and jumpahead().  For a crypto-grade rng, all but the first of these
will be a no-op.
"""

import os
import random
import re

__version__ = '1.3.10'
__version_date__ = '2019-03-19'

__all__ = [ \
    # constants, so to speak
    '__version__', '__version_date__',
    'MAX_INT16', 'MAX_INT32', 'MAX_INT64',
    'FILE_NAME_CHARS', 'FILE_NAME_STARTERS', 'FILE_NAME_OTHER_CHARS',

    # functions
    'valid_file_name',

    # classes
    'SimpleRNG', 'SystemRNG', 'SecureRNG', 'DataFile'
]

# we pray for constant folding - and we would prefer that these be const
MAX_INT16 = 65536
MAX_INT32 = 65536 * 65536
MAX_INT64 = 65536 * 65536 * 65536 * 65536

# characters we permit in file names, other than the first position
FILE_NAME_CHARS =  \
    r'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-.'

# patterns used in recognizing valid file names
FILE_NAME_STARTERS = \
    r'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_'
FILE_NAME_OTHER_CHARS =  \
    r'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_\-\.'

VALID_FILE_NAME_PAT = \
    r'^[' + FILE_NAME_STARTERS + '][' + FILE_NAME_OTHER_CHARS + ']*$'
VALID_FILE_NAME_RE = re.compile(VALID_FILE_NAME_PAT)


def valid_file_name(name):
    """ Return whether the name matches the regular expression. """
    match = VALID_FILE_NAME_RE.match(name)
    return match is not None

# -------------------------------------------------------------------


class DataFile(object):
    """ This appears to be a stub USED ONLY IN TESTING """

    def __init__(self, name, parent=None):
        self._name = name
        self._parent = parent

    @property
    def name(self):
        """ Return the name of the data file. """
        return self._name

    @property
    def path(self):
        """ Return a relative or absolute path to the data file. """
        if self._parent:
            pth = os.path.join(self._parent.path, self._name)
        else:
            pth = self._name
        return pth

    @property
    def parent(self):
        """ Return the name of the data file's parent. """
        return self._parent

    def __eq__(self, other):
        """ Return whether two data files are the same. """

        if self is other:
            return True
        if other is None or self.name != other.name:
            return False
        if self.parent and self.parent != other.parent:
            return False
        return self._parent == other.parent

# -------------------------------------------------------------------


class CommonFunc(object):
    """
    Parent class for RNG classes.

    This class contains convenience functions to be added to Random.
    """

    def random(self):
        """ Subclasses must override. """
        raise NotImplementedError

    def next_boolean(self):
        """ Return a quasi-random boolean value. """

        return self.random() >= 0.5

    def next_byte(self, max_=256):
        """ Return a quasi-random byte value between 0 and 255 inclusive. """

        if max_ < 1:
            max_ = 1
        elif max_ > 256:
            max_ = 256

        return int(max_ * self.random())

    def _rand_bytes(self, count):
        _ = self
        for _ in range(count):
            yield random.getrandbits(8)

    def next_bytes(self, buf):
        """
        buf is a bytearray.  Fill it with random bytes.

        This is the version for the Mersenne Twister.  SystemRNG and
        SecureRNG should override.
        """
        if buf is not None:
            count = len(buf)
            if count <= 64:
                val = bytearray(self._rand_bytes(count))
            else:
                val = bytearray(os.urandom(count))
            buf[:] = val

    def some_bytes(self, count):
        """ return a bytearray of N random bytes """
        buffer = bytearray(count)
        self.next_bytes(buffer)
        return buffer

    def next_int16(self, max_=65536):
        """ Return a quasi-random 16-bit int < max_. """

        if (max_ <= 0) or (max_ > 65536):
            max_ = 65536
        return int(max_ * self.random())

    def next_int32(self, max_=(65536 * 65536)):
        """ Return a quasi-random 32-bit int < max_. """
        if (max_ <= 0) or ((65536 * 65536) < max_):
            max_ = (65536 * 65536)
        return int(max_ * self.random())

    def next_int64(self, max_=(65536 * 65536 * 65536 * 65536)):
        """ Return a quasi-random 64-bit int < max_. """

        if (max_ <= 0) or ((65536 * 65536 * 65536 * 65536) < max_):
            max_ = (65536 * 65536 * 65536 * 65536)
        return int(max_ * self.random())

    def next_real(self):
        """
        Return a quasi-random floating-point number in the range [0..1).
        """

        return self.random()

    # ---------------------------------------------------------------

    # These produce strings which are acceptable POSIX file names
    # and also advance a cursor by a multiple of 64 bits.  All strings
    # are at least one byte and less than max_Len bytes n length.  We
    # arbitrarily limit file names to less than 256 characters.

    def _next_file_name(self, name_len):
        """ Always returns at least one character. """

        max_starter_ndx = len(FILE_NAME_STARTERS)
        ndx = self.next_byte(max_starter_ndx)
        name = FILE_NAME_STARTERS[ndx]
        max_char_ndx = len(FILE_NAME_CHARS)
        for _ in range(name_len - 1):
            ndx = self.next_byte(max_char_ndx)
            char = FILE_NAME_CHARS[ndx]
            name = name + char
        return name

    def next_file_name(self, max_len):
        """ Return a legal file name with 0 < length < max_len). """
        if max_len < 2:
            max_len = 2      # this is a ceiling which cannot be reached
        name_len = 0
        while name_len == 0:
            name_len = self.next_byte(max_len)     # so len < 256
        while True:
            name = self._next_file_name(name_len)
            if name and (name.find("..") == -1):
                break
        return name

    # These are operations on the file system.  Directory depth is at least 1
    # and no more than 'depth'.  Likewise for width, the number of
    # files in a directory, where a file is either a data file or a
    # subdirectory. The number of bytes in a file is at least min_len and
    # less than max_len.
    # Subdirectory names may be random

    def next_data_file(self, dir_name, max_len, min_len=0):
        """
        Return a data file in directory dir_name with a quasi-random name
        and contents.   The file is at least min_len bytes log and less than
        max_len bytes long.  Parameters are silently converted to reasonable
        values if necessary.
        """

        if min_len < 0:
            min_len = 0
        if max_len < min_len + 1:
            max_len = min_len + 1

        # loop until name does not match existing file
        path_to_file = "%s/%s" % (dir_name, self.next_file_name(16))
        while os.path.exists(path_to_file):
            path_to_file = "%s/%s" % (dir_name, self.next_file_name(16))

        count = min_len + int(self.random() * (max_len - min_len))
        data = self.some_bytes(count)

        with open(path_to_file, "wb") as file:
            file.write(data)
            # could check file size with file.tell()
        return (count, path_to_file)

    # BUGS
    # * on at least one occasion with width = 4 only 3 files/directories
    #   were created at the top level (2 were subdirs)
    # DEFICIENCIES:
    # * no control over percentage of directories
    # * no guarantee that depth will be reached
    def next_data_dir(self, path_to_dir, depth, width, max_len, min_len=0):
        """ Creates a directory tree populated with data files. """
        # number of directory levels; 1 means no subdirectories
        if depth < 1:
            depth = 1
        # number of members (files, subdirectories) at each level
        if width < 1:
            width = 1
        if not os.path.exists(path_to_dir):
            os.makedirs(path_to_dir)
        subdir_so_far = 0
        for i in range(width):
            if depth > 1:
                if (self.random() > 0.25) and (
                        (i < width - 1) or (subdir_so_far > 0)):
                    # 25% are subdirs
                    # data file i
                    # SPECIFICATION ERROR: file name may not be unique
                    (_, path_to_file) = self.next_data_file(
                        path_to_dir, max_len, min_len)
                    _ = path_to_file
                else:
                    # directory
                    subdir_so_far += 1
                    # create unique name
                    file_name = self.next_file_name(16)
                    path_to_subdir = os.path.join(path_to_dir, file_name)
                    self.next_data_dir(path_to_subdir, depth - 1, width,
                                       max_len, min_len)
            else:
                # data file
                # SPECIFICATION ERROR: file name may not be unique
                (_, path_to_leaf) = self.next_data_file(
                    path_to_dir, max_len, min_len)
                _ = path_to_leaf            # suppress warning ?


class SimpleRNG(random.Random, CommonFunc):
    """ if salt is None, uses time of day as salt """

    def __init__(self, salt=None):
        super().__init__(salt)    # in first parent


class SystemRNG(random.SystemRandom, CommonFunc):
    """
    A more secure random number generator getting numbers from the
    system's /dev/urandom.  This will be slower than SimpleRNG but
    not so very slow as an RNG using /dev/random, which will block
    until enough entropy accumulates.
    """

    def __init__(self, salt=None):
        super().__init__()      # "useless super delegation" ?
        _ = salt                # make pylint happy

    def getstate(self):
        """ Implements abstract function. """
        raise NotImplementedError('not implemented, stateless RNG')

    def setstate(self, state):
        """ Implements abstract function. """
        raise NotImplementedError('not implemented, stateless RNG')

    def next_byte(self, max_=256):
        """
        Return a quasi-random byte value between 0 and max_ - 1 inclusive.
        """

        if max_ < 1:
            max_ = 1
        elif max_ > 256:
            max_ = 256

        val = os.urandom(1)[0]
        if max_ < 256:
            val *= float(max_) / 256
        return int(val)

    def next_bytes(self, buf):
        """
        buf is a bytearray.  Fill it with random bytes.
        """
        if buf is not None:
            count = len(buf)
            buf[:] = bytearray(os.urandom(count))


class SecureRandom(random.Random):
    """
    Overrides Random.random(), stubs the other 5 functions.
    """

    BPF = 53        # bits in a Python float
    RECIP_BPF = 2 ** -BPF

    def __init__(self, salt=None):
        super().__init__()          # useless super delegation ?
        _ = salt                    # to suppress pylint complaints

    def _random(self, k):
        """ Read /dev/random for k bytes: blocks. """

        assert k >= 0
        with open('/dev/random', 'rb') as file:
            return file.read(k)

    def random(self):
        """ Return a random value in the range [0..1) """
        # DEBUG
        print("SecureRandom.random")
        # END
        return (int.from_bytes(self._random(7), 'little') >> 3) * \
            SecureRandom.RECIP_BPF

    # def seed(self):     # a=None, version=2):
    @staticmethod
    def seed(a=None, version=2):
        """ Unused abstract method. """
        # _,_ = a, version
        raise NotImplementedError('not implemented, stateless RNG')
        # return

    def jumpahead(self):
        """ Unused abstract method. """
        pass

    def getstate(self):
        """ Implements abstract function. """
        raise NotImplementedError('not implemented, stateless RNG')

    def setstate(self, state):
        """ Implements abstract function. """
        _ = state           # suppress warnings
        raise NotImplementedError('not implemented, stateless RNG')

    def _notimplemented(self):
        """ Implements abstract function. """
        raise NotImplementedError()


class SecureRNG(SecureRandom, CommonFunc):
    """
    SecureRandom plus the common functions,
    """

    def __init__(self, salt=0):
        super().__init__()    # in first parent, I hope
        # self.seed(salt)

    def _notimplemented(self):
        """ Implements abstract function. """
        raise NotImplementedError()

    def next_byte(self, max_=256):
        """
        Return a quasi-random byte value between 0 and max_ - 1 inclusive.
        """

        if max_ < 1:
            max_ = 1
        elif max_ > 256:
            max_ = 256

        val = self._random(1)[0]
        if max_ < 256:
            val *= float(max_) / 256
        return int(val)

    def next_bytes(self, buf):
        """
        buf is a bytearray.  Fill it with random bytes.
        """
        if buf is not None:
            count = len(buf)
            buf[:] = bytearray(self._random(count))
